test/core/primitives/primitives_codec_test.cpp

```diff
@@ -179,3 +271,186 @@ TEST_F(Primitives, decodeBlock) {
+  ASSERT_EQ(ver.implName(), version().implName());
+  ASSERT_EQ(ver.authoringVersion(), version().authoringVersion());
+  ASSERT_EQ(ver.apis(), version().apis());
+}
+
+/// BlockId
+
+/**
+ * @given block id constructed as hash256 and predefined match
+ * @when codec()->encodeBlockId() is applied
+ * @then obtained result matches predefined value
+ */
+TEST_F(Primitives, encodeBlockIdHash256) {
+  auto &&res = codec()->encodeBlockId(blockIdHash());
+  ASSERT_TRUE(res);
+  ASSERT_EQ(res.value(), blockIdHashMatch());
+}
+
+/**
+ * @given block id constructed as BlockNumber and predefined match
+ * @when codec()->encodeBlockId() is applied
+ * @then obtained result matches predefined value
+ */
+TEST_F(Primitives, encodeBlockIdBlockNumber) {
+  auto &&res = codec()->encodeBlockId(blockIdNumber());
+  ASSERT_TRUE(res);
+  ASSERT_EQ(res.value(), (blockIdNumberMatch()));
+}
+
+/**
+ * @given byte array containing encoded block id
+ * and predefined block id as Hash256
+ * @when codec()->decodeBlockId() is applied
+ * @thenobtained result matches predefined block id instance
+ */
+TEST_F(Primitives, decodeBlockIdHash) {
+  auto stream = ByteArrayStream(blockIdHashMatch());
+  auto &&res = codec()->decodeBlockId(stream);
+  ASSERT_TRUE(res);
+  ASSERT_EQ(res.value(), blockIdHash());
+}
+
+/**
+ * @given byte array containing encoded block id
+ * and predefined block id as BlockNumber
+ * @when codec()->decodeBlockId() is applied
+ * @thenobtained result matches predefined block id instance
+ */
+TEST_F(Primitives, decodeBlockIdNumber) {
+  auto stream = ByteArrayStream(blockIdNumberMatch());
+  auto &&res = codec()->decodeBlockId(stream);
+  ASSERT_TRUE(res);
+  ASSERT_EQ(res.value(), blockIdNumber());
+}
+
+/// TransactionValidity
+
+/**
+ * @given TransactionValidity instance as Invalid and predefined match
+ * @when codec()->encodeTransactionValidity() is applied
+ * @then obtained result matches predefined value
+ */
+TEST_F(Primitives, encodeTransactionValidityInvalid) {
+  TransactionValidity invalid = Invalid{1};
+  auto &&res = codec()->encodeTransactionValidity(invalid);
+  ASSERT_TRUE(res);
+  ASSERT_EQ(res.value(), (Buffer{0, 1}));
+}
+
+/**
+ * @given TransactionValidity instance as Unknown and predefined match
+ * @when codec()->encodeTransactionValidity() is applied
+ * @then obtained result matches predefined value
+ */
+TEST_F(Primitives, encodeTransactionValidityUnknown) {
+  TransactionValidity unknown = Unknown{2};
+  auto &&res = codec()->encodeTransactionValidity(unknown);
+  ASSERT_TRUE(res);
+  ASSERT_EQ(res.value(), (Buffer{2, 2}));
+}
+
+/**
+ * @given TransactionValidity instance as Valid and predefined match
+ * @when codec()->encodeTransactionValidity() is applied
+ * @then obtained result matches predefined value
+ */
+TEST_F(Primitives, encodeTransactionValidity) {
+  auto &&res = codec()->encodeTransactionValidity(validTransaction());
+  ASSERT_TRUE(res);
+  ASSERT_EQ(res.value(), validMatch());
+}
+
+namespace {
```

Comment:
Just use visis_in_place

Reply:
migrating to boost::variant in progress...

Reply:
migration complete, progress 100% )